Add `tbot proxy` and `tbot db` wrapper commands

timothyb89 opened the pull request:
This adds new wrapper commands that leverage tsh for proxy and database access.

It also adds a new `tshwrap` helper package which contains utilities for locating the tsh executable, checking its version, and loading all necessary data (certificates, destinations, etc) that will need to be passed to tsh for wrapped commands to function.

Usage examples:
* SSH: `ssh -F tbot-user/ssh_config user@host.example.com`
   (should now work both with and without the legacy SSH proxy port enabled)
* Database connect: `tbot db --destination-dir=./tbot-user --proxy=proxy.example.com connect db-name`
* Database proxy: `tbot proxy --destination-dir=./tbot-user --proxy=proxy.example.com:3080 db db-name`
* Database proxy with authenticated tunnel: `tbot proxy --destination-dir=./tbot-user --proxy=proxy.example.com:3080 --tunnel db-name`

Fixes #11445

timothyb89 committed:
Extend support for identity files in tsh

This enhances support for identity files in tsh, which previously only
worked for regular SSH access. The largest blocker for support is that
tsh uses profiles for all non-SSH resource access, and profiles have a
direct mapping to some on-disk resources. This patch works around this
in a few ways:
 * Virtual profiles: When an identity file is specified with `-i`, we
   use it to create an in-memory virtual profile using the cert as the
   root identity _and_ for every `RouteToDatabase` (and in the future,
   app) field contained in the cert.
 * Virtual profile paths: Certain profile operations require paths to
   valid certificates and other resources on disk, which may not exist
   inside the identity file.

   As the driving use-case for this change is integration with Machine
   ID, we can "cheat" and pass the correct paths to tsh via
   environment variables. A cooperating wrapper in `tbot` will execute
   `tsh` with appropriate flags and environment variables, which
   override tsh's usual certifiate paths. This allows commands like
   `tsh db connect ...` to work as expected.
 * Key stores: previously we used a `noLocalKeyStore{}` with which all
   lookups fail. This patch replaces it with an in-memory keystore if
   a client key is available.
 * Profile status: lastly, we add a new `StatusCurrentWithIdentity()`
   function to load virtual profiles where supported. Some commands
   are not supported in this PR (like `tsh app ...`), but others
   don't make sense to support (like cert reissuing).

   We might consider merging everything into the traditional
   `StatusCurrent()` when adding app support.

App access is still broken and will be addressed in a later change.

Partially fixes #11770

timothyb89 committed:
Add `tbot proxy` and `tbot db` wrapper commands

This adds new wrapper commands that leverage tsh for proxy and
database access.

It also adds a new `tshwrap` helper package which contains utilities
for locating the tsh executable, checking its version, and loading
all necessary data (certificates, destinations, etc) that will need
to be passed to tsh for wrapped commands to function.

timothyb89 commented:
Outstanding TODOs:

- [ ] Merge #12686 
- [ ] Use `tbot proxy ssh` in generated `ssh_config`

timothyb89 committed:
Fix failing lint

timothyb89 committed:
Fix failing unit test due to incorrect default IsVirtual profile flag

timothyb89 committed:
Combine `StatusCurrentWithIdentity()` into `StatusCurrent()`

Additionally, log a warning when environment variable paths aren't
found.

timothyb89 committed:
Fix virtual profile flag always being true

timothyb89 committed:
Merge remote-tracking branch 'origin/master' into timothyb89/tsh-expanded-identityfiles

timothyb89 committed:
Update lib/client/api.go

Co-authored-by: Krzysztof Skrzętnicki <krzysztof.skrzetnicki@goteleport.com>

timothyb89 committed:
Address review feedback

timothyb89 committed:
Merge remote-tracking branch 'origin/master' into timothyb89/tsh-expanded-identityfiles

timothyb89 committed:
Merge branch 'timothyb89/tsh-expanded-identityfiles' into timothyb89/tbot-proxy-wrappers

timothyb89 committed:
Use `tbot proxy` in generated `ssh_config`

timothyb89 committed:
Merge branch 'master' into timothyb89/tbot-proxy-wrappers

timothyb89 committed:
Add tests for mockable parts of our tsh integration

timothyb89 committed:
Merge remote-tracking branch 'origin/master' into timothyb89/tbot-proxy-wrappers

timothyb89 committed:
Merge remote-tracking branch 'origin/timothyb89/tbot-proxy-wrappers' into timothyb89/tbot-proxy-wrappers

timothyb89 committed:
Fix lints

timothyb89 committed:
Clarify docstrings in CLIConf

strideynet commented on tool/tbot/config/configtemplate_ssh_client.go:
```diff
@@ -134,6 +135,10 @@ func (c *TemplateSSHClient) Describe() []FileDescription {
+// using non-filesystem backends.
+var sshConfigUnsupportedWarning sync.Once
+
 func (c *TemplateSSHClient) Render(ctx context.Context, authClient auth.ClientI, currentIdentity *identity.Identity, destination *DestinationConfig) error {
```
It would be super nice to include a test for this function using the `Golden` helper, so we can easily see what the rendered file looks like in diffs

strideynet replied:
I've gone ahead and done this in https://github.com/gravitational/teleport/pull/12898 . It should be simple enough to pull that commit into this branch, or approve and merge that PR and then rebase onto it.

You'll need to run `GOLDEN_UPDATE=1 go test -timeout 10m -run ^TestTemplateSSHClient_Render$ github.com/gravitational/teleport/tool/tbot/config` to regenerate the golden file for your changes in this PR.

timothyb89 replied:
I am a big fan of making the result of these changes obvious during review. Any objections if we try to get this in after v9.3?

timothyb89 replied:
Ah, yep, I've fixed the failing test, making `ssh_config` optional broke it.

I do like the golden approach but let's aim to include that post-v9.3.

strideynet reviewed (approved):
This looks solid. There's a test failure for `TestDefaultTemplateRendering` which may be relevant though.

It might be easier to rebase this on top of https://github.com/gravitational/teleport/pull/12855 when thats been merged rather than vice versa.

strideynet commented on tool/tbot/tshwrap/wrap.go:
```diff
@@ -0,0 +1,311 @@
+}
+
+// GetDestinationPath returns a path to a filesystem destination.
+func GetDestinationPath(destination *config.DestinationConfig) (string, error) {
```
It would probably be quite easy to include a unit test for this.

strideynet commented on tool/tbot/tshwrap/wrap.go:
```diff
@@ -0,0 +1,311 @@
+
+// GetTLSCATemplate returns the TLS CA template for the given destination. It's
+// a required template so this should never fail.
+func GetTLSCATemplate(destination *config.DestinationConfig) (*config.TemplateTLSCAs, error) {
```
It would probably be quite easy to include a unit test for this.

strideynet commented on tool/tbot/tshwrap/wrap.go:
```diff
@@ -0,0 +1,311 @@
+
+// GetIdentityTemplate returns the identity template for the given destination.
+// This is a required template so it _should_ never fail.
+func GetIdentityTemplate(destination *config.DestinationConfig) (*config.TemplateIdentity, error) {
```
It would probably be quite easy to include a unit test for this.

timothyb89 commented:
> It might be easier to rebase this on top of #12855 when thats been merged rather than vice versa.

A caveat here is that we are looking to squeeze this into v9.3 (end of this week). I have mixed feelings about also squeezing in and backporting a refactoring patch, especially with the timeline constraints. What do you think?

timothyb89 commented on tool/tbot/proxy.go:
```diff
@@ -0,0 +1,73 @@
+	"github.com/gravitational/trace"
+)
+
+func onProxyDBCommand(botConfig *config.BotConfig, cf *config.CLIConf) error {
```
...oh, yeah, that's a typo. We want to wrap the entire `tsh proxy` subcommand, including `db` and `ssh`. We'll be taking advantage of `app` later so I'm okay leaving it quasi-supported.

I'll rename the function name, it's a holdover from before I'd determined we could wrap the whole command without splitting out handlers for each type.

strideynet commented:
> > It might be easier to rebase this on top of #12855 when thats been merged rather than vice versa.
> 
> A caveat here is that we are looking to squeeze this into v9.3 (end of this week). I have mixed feelings about also squeezing in and backporting a refactoring patch, especially with the timeline constraints. What do you think?

If that's the case, just get that into 9.3 :)

timothyb89 committed:
Tweak comment for clarity; fix typo in `onProxyCommand`

timothyb89 committed:
Add missing copyright header

timothyb89 committed:
Fix failing unit test and pass destination to `Describe()`

This fixes a failing unit test by making the description for
`ssh_config` match its behavior in practice. This necessitated
passing the destination to all templates, unfortunately.

timothyb89 committed:
Add a few extra tests

timothyb89 commented on tool/tbot/main.go:
```diff
@@ -98,13 +98,39 @@ func Run(args []string, stdout io.Writer) error {
 	}
 
+	// It's unfortunate but we'll need to handle remaining args after parsing
+	// to ensure we don't overwrite anything.
```
I've rewritten the comment to hopefully reduce its word salad-ness.

codingllama commented on tool/tbot/tshwrap/wrap_test.go:
```diff
@@ -0,0 +1,191 @@
+	captureOut []byte
+}
+
+func (r *mockRunner) Capture(args ...string) ([]byte, error) {
```
I expect this mock is going to be one of the counter-arguments to my "kill the interface" comment, so I'd like to present alternatives.

1. Mock exec functions

This is similar to how folks mock `time.Now()`. You do a `var timeNow = time.Now` and use that var in your prod code. Tests swap the value for a mocked function (reset via t.Cleanup).

In your case, it's likely exec.Command, so it would be a bit of work. Other alternatives are better.

2. Detach Capture from the interface and mock it.

Ie, `var capture = func(path string, args []string) {...}` (or something like that). Same deal as above from here on.

3. Capture is mostly (only?) used for "tsh version", so create (and later mock!) a more specialized func for "tsh version". 

Possibly `var tshVersion = func(path string) (string, error) {...}`. The exec.Command invocation lives inside. Returns the JSON string on success.


These are common techniques in Go that, I believe, would push your API to be a little more idiomatic.

codingllama commented on tool/tbot/tshwrap/wrap_test.go:
```diff
@@ -0,0 +1,191 @@
+// implicitly, that the version capture and parsing works.)
+func TestTSHSupported(t *testing.T) {
+	version := func(v string) []byte {
+		bytes, err := json.Marshal(struct {
```
Just write the string here?

codingllama commented on tool/tbot/tshwrap/wrap_test.go:
```diff
@@ -0,0 +1,191 @@
+		name   string
+		out    []byte
+		err    error
+		expect func(t *testing.T, err error)
```
Can you mimic require.Error's signature here so you can pass it directly to this field?

timothyb89 replied:
ah, yep, that's a nice simplification, thanks!

codingllama commented on tool/tbot/tshwrap/wrap_test.go:
```diff
@@ -0,0 +1,191 @@
+
+	tests := []struct {
+		name   string
+		out    []byte
+		err    error
```
```suggestion
		mock *mockRunner
```

instead of dismembering the runner? Makes it easier to reason about what happens with out/err.

codingllama commented on tool/tbot/tshwrap/wrap_test.go:
```diff
@@ -0,0 +1,191 @@
+	env, err := GetEnvForTSH(&destination)
+	require.NoError(t, err)
+	for k, v := range expected {
+		require.Contains(t, env, k)
+		require.Equal(t, v, env[k])
```
Simplify? Also let all errors accumulate?

```suggestion
		assert.Equal(t, v, env[k])
```

codingllama reviewed (commented):
Apologies for the delay.

timothyb89 commented on tool/tbot/config/configtemplate_cockroach.go:
```diff
@@ -22,6 +22,7 @@ import (
 	"github.com/gravitational/teleport/api/types"
```
Luckily @strideynet is working on that already in https://github.com/gravitational/teleport/pull/12855 :)

timothyb89 commented on tool/tbot/config/configtemplate_ssh_client.go:
```diff
@@ -155,36 +166,44 @@ func (c *TemplateSSHClient) Render(ctx context.Context, authClient auth.ClientI,
+	// writing ssh_config. (Future users of k8s secrets will need to bring
+	// their own config, we can't predict where paths will be in practice.)
+	if destDir == "" {
+		sshConfigUnsupportedWarning.Do(func() {
+			log.Infof("Note: no ssh_config will be written for non-filesystem "+
```
I think so, it's not entirely actionable, just expected behavior if using a different destination backend (memory today, k8s secret / etc in the future). The idea is to preemptively address confusion around "this file I expected to show up doesn't show up in this circumstance"

timothyb89 committed:
Apply suggestions from code review

Co-authored-by: Alan Parra <alan.parra@goteleport.com>

timothyb89 commented on tool/tbot/tshwrap/wrap.go:
```diff
@@ -0,0 +1,327 @@
+}
+
+// GetTLSCATemplate returns the TLS CA template for the given destination. It's
+// a required template so this should never fail.
```
always :) but that's why we have tests in place for it

codingllama replied:
![image](https://user-images.githubusercontent.com/12500300/170716813-c67c4a42-56b8-40e8-8829-33a48ffdedbd.png)

timothyb89 commented on tool/tbot/tshwrap/wrap.go:
```diff
@@ -0,0 +1,327 @@
+	// The subprocess should inherit the environment plus our vars.
+	environ := os.Environ()
+	for k, v := range env {
+		environ = append(environ, k+"="+v)
```
Per [the docs](https://pkg.go.dev/os/exec#Cmd), yes, the last one wins. It's bonkers to me this isn't a map but at least it handles lists sensibly.

codingllama replied:
Yep, this is a strange one.  ¯\_(ツ)_/¯

timothyb89 commented on tool/tbot/main.go:
```diff
@@ -98,13 +98,40 @@ func Run(args []string, stdout io.Writer) error {
+	// Remaining args are stored directly to a []string rather than written to
+	// a shared ref like most other kingpin args, so we'll need to manually
+	// move them to the remaining args field.
+	if len(*dbRemaining) > 0 {
```
I don't think kingpin allows for that unfortunately. This is ugly but I think it's the best we can get :disappointed:

codingllama replied:
Kingpin is a bit of a strange choice, [cobra](https://github.com/spf13/cobra) is much more common nowadays (I think it leads to better code too, no need for that "what's the command" switch). I assume Teleport predates cobra, hence kingpin.

Just idle musings, no actions expected here.

timothyb89 commented on tool/tbot/proxy.go:
```diff
@@ -0,0 +1,73 @@
+
+	// Pass through the debug flag.
+	if botConfig.Debug {
+		args = append([]string{"-d"}, args...)
```
Unfortunately it won't work here, `-d` cannot come after the `proxy` subcommand arg. We could split the initial `args` definition but at that point we're deciding between uglies and I'd rather keep it as written.

timothyb89 committed:
Address another batch of review comments

timothyb89 commented on tool/tbot/config/configtemplate.go:
```diff
@@ -88,7 +88,7 @@ type Template interface {
 	// secrets) for use with `tbot init`. If an arbitrary number of files must
 	// be generated, they should be placed in a subdirectory.
-	Describe() []FileDescription
+	Describe(destination destination.Destination) []FileDescription
```
It's come up before and I agree that it's ugly, but this practically needs to live in its own package to avoid import cycles. Since it's the only thing that will ever live in that package, there's few names that would make sense other than, well, `destination`. Always open to suggestions if you've got a more creative solution to offer :slightly_smiling_face:

codingllama replied:
Having a quick look the only implementor I found was DestinationDirectory. The `destination` package holds an interface and nothing else, which is a bit of a design smell for Go.

I would suggest the following:

* Rename `package destination` to `package destinations`. The main reason is that we can have variables named `destination` without overshadowing the package name (similar to `strings` or `bytes`).
* Move `DestinationDirectory` to `destinations.Directory`. Names fit nicely and the package now exports something concrete. (IIUC, this won't cause import cycles.)
* Ditch the interface, or at the very least ditch it from the `destinations` package. If you do need an interface, spring up one where it's used, preferably with only 1 or 2 methods that the code cares about.

We don't usually need "preemptive" interfaces in Go. Code against the concrete type if you can; if you need an interface, spit it out where it's needed, and only after it has a reason to be there.  (The api/ package is the one place where we may want to do differently, but for internal code this should work just fine.)

timothyb89 commented on tool/tbot/config/configtemplate_ssh_client.go:
```diff
@@ -123,17 +125,26 @@ func (c *TemplateSSHClient) Name() string {
 		},
 	}
+
+	if _, ok := destination.(*DestinationDirectory); ok {
```
You're right, this deserves a comment. It is mentioned down below in `Render()` where it's actually (not) written but I'll make a note up here as well.

timothyb89 commented on tool/tbot/db.go:
```diff
@@ -0,0 +1,80 @@
+	args = append(args, cf.RemainingArgs...)
+
+	// Pass through the debug flag.
+	if botConfig.Debug {
```
Unfortunately we can't include `-d` after `db`. There's other ways we could satisfy all our argument ordering constraints but they all end up pretty ugly themselves. In the end I think as ugly as this is, it's a wash.

timothyb89 committed:
Comment tweaks

timothyb89 commented on tool/tbot/tshwrap/wrap.go:
```diff
@@ -0,0 +1,327 @@
+// AnyArgsStartWith determines if any of the string arguments have the given
+// prefix. Useful for determining if an arg for tsh got caught in
+// RemainingArgs.
+func AnyArgsStartWith(prefix string, args []string) bool {
```
For better or worse I've moved it to `lib/utils`

codingllama replied:
I share the sentiment. Probably for the best, though.

timothyb89 commented on tool/tbot/proxy.go:
```diff
@@ -0,0 +1,73 @@
+		return trace.Wrap(err)
+	}
+
+	identityPath := filepath.Join(destinationPath, identityTemplate.FileName)
```
I do like moving to a simple `Wrapper` struct returned by `New()`. I'm wary of coupling all the extra fields so tightly because ultimately beyond just the path to tsh itself, all arguments to tsh are determined by the caller. The cost of gathering everything every time even if we don't need it is minimal, but then so is extracting just what we need given there's already helper functions.

Ultimately the difference feels pretty arbitrary and I'd rather err on the side of leaving things that aren't broken.

timothyb89 commented on tool/tbot/tshwrap/wrap.go:
```diff
@@ -0,0 +1,327 @@
+}
+
+// GetTSHVersion queries the system tsh for its version.
+func GetTSHVersion(r TSHRunner) (*semver.Version, error) {
```
I think it makes some sense to keep this uncoupled. The only other wrapper command I (maybe) anticipate is `tsh ssh`, which has no version dependency at all, meaning it wouldn't need to call `tsh version` and risk erroring out on a pre-`-f json` tsh that would otherwise work fine.

codingllama replied:
Fair enough.

timothyb89 committed:
Refactor tshwrap to remove the Runner interface.

codingllama commented on tool/tbot/db.go:
```diff
@@ -0,0 +1,82 @@
+	}
+	args = append(args, cf.RemainingArgs...)
+
+	// Pass through the debug flag, and prepend to satisfy argument ordering
+	// needs.
```
```suggestion
	// Pass through the debug flag, and prepend to satisfy argument ordering
	// needs (has to come after proxy flags).
```

Same for the others (assuming I got the reason correctly, feel free to rephrase!)

codingllama commented on tool/tbot/tshwrap/wrap.go:
```diff
@@ -0,0 +1,327 @@
+const (
+	// TSHVarName is the name of the environment variable that can override the
+	// tsh path that would otherwise be located on the $PATH.
+	TSHVarName = "TSH"
```
That's fair. Maybe `tshwrap.EnvVarName` would cut it, but it's fine as-is.

codingllama commented on tool/tbot/tshwrap/wrap.go:
```diff
@@ -0,0 +1,317 @@
+	trace.Component: teleport.ComponentTBot,
+})
+
+// TSHRunner is a method for executing tsh.
+type TSHRunner interface {
+	// capture runs tsh and captures its standard output.
+	capture(args ...string) ([]byte, error)
+
+	// Exec runs tsh with the given additional environment variables and args.
+	// It should inherit the current process's input and output streams.
+	Exec(env map[string]string, args ...string) error
+}
+
```
```suggestion

```

dead code?

timothyb89 replied:
whoops, yup!

codingllama reviewed (approved):
Just a few ramblings / minor suggestions. Thanks for the fixes.

timothyb89 committed:
Apply suggestions from code review

Co-authored-by: Alan Parra <alan.parra@goteleport.com>

timothyb89 committed:
Address review comments

timothyb89 committed:
Merge branch 'master' into timothyb89/tbot-proxy-wrappers
